### PR TITLE
[ws-daemon] Refactor CPU limiting

### LINKED ISSUE #8035: Implement globally optimal dynamic CPU limiting in ws-daemon

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR refactors the dynamic CPU limiting performed by ws-daemon. More details regarding motivation and solution can be found in this loom video: https://www.loom.com/share/0922bf11430c4ac084ebab6616e40b7e

This PR also adds a set of metrics to observe the new CPU limits:
- `gitpod_ws_daemon_cpulimit_workspaces_added_total`: Number of workspaces added to CPU control
- `gitpod_ws_daemon_cpulimit_workspaces_removed_total` Number of workspaces removed from CPU control
- `gitpod_ws_daemon_cpulimit_workspaces_burst_total`: Number of workspaces which received burst - `gitpod_ws_daemon_cpulimit_workspaces_cputime_seconds` CPU time of all observed workspaces
- `gitpod_ws_daemon_cpulimit_workspaces_throttled_total` Number of workspaces which ran with throttled CPU

## Related Issue(s)

Fixes #8035

## How to test
By default dynamic CPU limiting is disabled. It can be enabled using the experimental config in the installer, or by modifying the ws-daemon configmap. Therein, set the cpuLimiting.enabled to true, and configure the total available bandwidth on the node (e.g. 12 CPUs), limit and burst limit.

Then, start a few workspaces, observe their `/sys/fs/cgroup/cpu/cpu.cfs_quota_us` value while producing some CPU load.

In addition, one could try this version on a single VM and experiment with workspaces exhibiting different load profiles.

## Chaveats
- this PR removes the process priority management we had prior. It's unclear how much that actually helped, considering that it very outdated and focused on `bash` and `theia` and `gitpod-node` only. If we find user experience deteriorated too much by this removal, we'll want to bring it back in a future change.
- this change does not yet introduce QoS classes, but that would be straight forward to add down the road using annotations on the pod.
- this change removes the fixed CPU limiting that was in place as agent-smith action (limit CPU), and that was meant to afford paying customers more CPU. It's doubtful that the latter had much of an effect, but we'll want to follow up quickly with the QoS change.

## Release Notes

```release-note
Refactor dynamic CPU limiting to provide fairer scheduling.
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** and additionally assign **mrsimonemms** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @mrsimonemms` in a comment when ready.

Associated issue: *#8035*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/common-go/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/common-go/OWNERS)**
- **[components/ws-daemon/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**
- **[installer/pkg/components/ws-daemon/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/pkg/components/ws-daemon/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/ws-daemon/pkg/cpulimit/cfs.go
Comment (utam0k) [inline components/ws-daemon/pkg/cpulimit/cfs.go:29]: I'm sorry if you have a plan to add handling it. We need to pay attention to zero-valued because a zero-valued cgroup value sometimes has a special meaning. For example, in `cpuacct.usage` zero means reset.
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/resource_management_guide/sec-cpuacct

Comment (csweichel) [reply components/ws-daemon/pkg/cpulimit/cfs.go:29]: Great point - thanks for pointing that out

Comment (utam0k) [reply components/ws-daemon/pkg/cpulimit/cfs.go:29]: Maybe the code of runc is helpful.
https://github.com/opencontainers/runc/blob/ede712786ce84d8e9953b694149428fc9dbe0d74/libcontainer/cgroups/fscommon/utils.go#L101

Comment (csweichel) [reply components/ws-daemon/pkg/cpulimit/cfs.go:30]: Thanks - the pointers were helpful indeed :)

**Review by corneliusludmann — DISMISSED** [review]
Installer changes look good to me.

/approve

File: components/ws-daemon/pkg/cpulimit/cfs.go
Comment (sagor999) [reply components/ws-daemon/pkg/cpulimit/cfs.go:30]: @csweichel just following up on above ^
seems like you need  to improve handling here?
should probably use `readUint64` function

File: components/ws-daemon/pkg/cpulimit/cfs.go
Comment (sagor999) [inline components/ws-daemon/pkg/cpulimit/cfs.go:27]: this should probably say that cannot read `cpuacct.usage`

File: components/ws-daemon/pkg/cpulimit/cfs.go
Comment (sagor999) [inline components/ws-daemon/pkg/cpulimit/cfs.go:49]: should say `cannot parse CFS quota`

File: components/ws-daemon/pkg/cpulimit/cfs.go
Comment (sagor999) [inline components/ws-daemon/pkg/cpulimit/cfs.go:103]: can you add `l` to error output here so that it will be easier to debug later on if that call ever fails?

File: components/ws-daemon/pkg/cpulimit/cpulimit.go
Comment (sagor999) [inline components/ws-daemon/pkg/cpulimit/cpulimit.go:40]: should we panic here? can this be an error instead? panic would mean a hard crash, which might not be great for ws-daemon

File: components/ws-daemon/pkg/cpulimit/cpulimit.go
Comment (sagor999) [inline components/ws-daemon/pkg/cpulimit/cpulimit.go:8]: I think you already added metrics?

**Comment by csweichel** [comment]
@sagor999 I've incorporated your feedback - please take another look

**Review by sagor999 — APPROVED** [review]

**Review by corneliusludmann — APPROVED** [review]

**Pull request merged**
